src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -227,34 +253,174 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+    // - delay: time in milliseconds to wait before performing the search
+    // (grace time to allow next search to start)
     // Return Value:
     // - <none>
-    void TermControl::_Search(const winrt::hstring& text, const bool goForward, const bool caseSensitive)
+    fire_and_forget TermControl::_SearchAsync(std::optional<bool> goForward, Windows::Foundation::TimeSpan const& delay)
     {
-        if (text.size() == 0 || _closing)
+        // Run only if the live search state was initialized
+        if (_closing || !_searchState.has_value())
         {
             return;
         }
 
-        const Search::Direction direction = goForward ?
-                                                Search::Direction::Forward :
-                                                Search::Direction::Backward;
+        const auto originalSearchId = _searchState.value().SearchId;
+        auto weakThis{ this->get_weak() };
 
-        const Search::Sensitivity sensitivity = caseSensitive ?
-                                                    Search::Sensitivity::CaseSensitive :
-                                                    Search::Sensitivity::CaseInsensitive;
+        // If no matches were computed it means we need to perform the search
+        if (!_searchState.value().Matches.has_value())
+        {
+            // Before we search, let's wait a bit:
+            // probably the search criteria or the data are still modified.
+            co_await winrt::resume_after(delay);
 
-        Search search(*GetUiaData(), text.c_str(), direction, sensitivity);
-        auto lock = _terminal->LockForWriting();
-        if (search.FindNext())
+            // Switch back to Dispatcher so we can set the Searching status
+            co_await winrt::resume_foreground(Dispatcher());
+            if (auto control{ weakThis.get() })
+            {
+                // If search box was collapsed or the new one search was triggered - let's cancel this one
+                if (!_searchState.has_value() || _searchState.value().SearchId != originalSearchId)
+                {
+                    co_return;
+                }
+
+                // Let's mark the start of searching
+                if (_searchBox)
+                {
+                    _searchBox->SetStatus(_searchState.value().Status());
+                    _searchBox->SetNavigationEnabled(false);
+                }
+
+                std::vector<std::pair<COORD, COORD>> matches;
+                if (!_searchState.value().Text.empty())
+                {
+                    const Search::Sensitivity sensitivity = _searchState.value().CaseSensitive ?
+                                                                Search::Sensitivity::CaseSensitive :
+                                                                Search::Sensitivity::CaseInsensitive;
+
+                    Search search(*GetUiaData(), _searchState.value().Text.c_str(), Search::Direction::Forward, sensitivity);

```

Comment:
We need to collect all results here, and then use them for both going backward and forward. Since when going forward we want to see `1 of n` and when going backward we want to see `n of n`, `n - 1 of n`,  the search should be done forward.

Reply:
Oh! You added a comment! Thanks!